### DISCUSSION TITLE
FSM

### DISCUSSION THREAD

**Discussion by miruken1207** [reo7sp/tgbot-cpp#306] [Q&A] [answered]
Hello everyone.
Who knows, is there realized fsm as in aiogram, cause I didn't find in documentation.
I am practicing C++ and wanted create bot for signing up and that is impossible without some kinda states.
Please, help me if you are able )

**Comment by reo7sp** [comment, marked as answer]
tgbot-cpp does not include an FSM like aiogram. Keep the state in the application, keyed by chat or user ID:

```cpp
enum class State { Idle, WaitingForName };
std::unordered_map<std::int64_t, State> states;

bot.getEvents().onCommand("signup", [&](TgBot::Message::Ptr message) {
    states[message->chat->id] = State::WaitingForName;
    bot.getApi().sendMessage(message->chat->id, "Enter your name");
});

bot.getEvents().onNonCommandMessage([&](TgBot::Message::Ptr message) {
    if (states[message->chat->id] != State::WaitingForName) {
        return;
    }
    const auto name = message->text.value_or("");
    states.erase(message->chat->id);
    bot.getApi().sendMessage(message->chat->id, "Registered: " + name);
});
```

Use a database if the state must survive restarts. Protect the map with a mutex when updates can be handled concurrently, for example with a webhook server.
